### PR TITLE
set default enabled admission plugins by official document

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

https://kubernetes.io/docs/admin/admission-controllers/#is-there-a-recommended-set-of-admission-controllers-to-use

recommend  running the following set of admission controllers 
```
If you previously had not set the `--admission-control` flag, your cluster behavior may change (to be more standard).  See [https://kubernetes.io/docs/admin/admission-controllers/] for explanation of admission control.
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Set default enabled admission plugins `NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota`
```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
And after this , the document also need revise.

File: pkg/kubeapiserver/options/plugins.go [hunk 1]
```diff
@@ -130,8 +130,17 @@ func RegisterAllAdmissionPlugins(plugins *admission.Plugins) {
 
 // DefaultOffAdmissionPlugins get admission plugins off by default for kube-apiserver.
 func DefaultOffAdmissionPlugins() sets.String {
-	defaultOffPlugins := sets.NewString(AllOrderedPlugins...)
-	defaultOffPlugins.Delete(lifecycle.PluginName)
+	defaultOnPlugins := sets.NewString(
+		lifecycle.PluginName,
```
Comment (deads2k) [inline pkg/kubeapiserver/options/plugins.go]: Alias the packages to admission plugin names so we can recognize these.  I think this is `namespacelifecycle`.

File: staging/src/k8s.io/apiserver/pkg/server/options/admission.go [hunk 2]
```diff
@@ -79,7 +79,7 @@ func NewAdmissionOptions() *AdmissionOptions {
 		// after all the mutating ones, so their relative order in this list
 		// doesn't matter.
 		RecommendedPluginOrder: []string{lifecycle.PluginName, initialization.PluginName, mutatingwebhook.PluginName, validatingwebhook.PluginName},
-		DefaultOffPlugins:      sets.NewString(initialization.PluginName, mutatingwebhook.PluginName, validatingwebhook.PluginName),
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/admission.go]: split this out, it is non-contentious

**Comment by deads2k** [comment]
This is a huge change.  I'm strongly in favor, but we need to agree that we're willing to do this.  

 1. People who previously controlled the admission chain using `--admission-control` will be fine and their options are still respected.
 2. People who specified no admission chains had a scarily exposed cluster they could not run securely, but now they'll suddenly get the admission chain and potentially breaks as things like service accounts will now be require.  We've long recommended this, but we've never turned them on by default before.

@smarterclayton @bgrant0607 @kubernetes/api-approvers @lavalamp @erictune @liggitt for comment on having a default admission chain.

File: pkg/kubeapiserver/options/plugins.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/kubeapiserver/options/plugins.go]: Add later.

**Comment by hzxuzhonghu** [comment]
ping @smarterclayton @bgrant0607 @kubernetes/api-approvers @lavalamp @erictune @liggitt for comment on having a default admission chain.

**Comment by lavalamp** [comment]
I am in favor of fixing this but I'm not sure this is the right way to do it.
* I think the UX is super confusing. It's not good to break naive users (e.g., I bet this breaks the local-up script). So, at a minimum I think we need users to opt-in to the new system.
* Changing entries in a default-on list is always going to be a breaking change. I think the right way to do this is to have apiserver store the list, probably somewhere in etcd. If the list is unset, then we write the defaults; otherwise, we use the list. That way, we could change the defaults without totally breaking everyone on upgrades.

**Comment by deads2k** [comment]
> I think the UX is super confusing. It's not good to break naive users (e.g., I bet this breaks the local-up script). So, at a minimum I think we need users to opt-in to the new system.

I'm ok with having a `--disable-admission` flag (which is what this would be).  I'd like to have a timeline for having `--disable-admission` default to false.

> Changing entries in a default-on list is always going to be a breaking change. I think the right way to do this is to have apiserver store the list, probably somewhere in etcd. If the list is unset, then we write the defaults; otherwise, we use the list. That way, we could change the defaults without totally breaking everyone on upgrades.

The idea of a "default on" list is that we think these are a good default for a safe and sane cluster, it's not a guarantee of particular plugins being on or off.  Think about things like the PodToleration plugin.  If you don't enable it in the same release that you added taint and toleration support, the feature presents 
 a risk to your cluster.  Most cluster-admins probably want to have their safe cluster continue being safe.

**Comment by sttts** [comment]
How many (non-toy/dev) clusters are out there without `--admission-control`?

Compared with the big switch to RBAC-by-default, any serious cluster probably has admission plugins enabled anyway. This was very different for RBAC:

**Comment by bgrant0607** [comment]
We need to think carefully about what we're trying to achieve.

If we want to make it easier for users to stay up to date with recommended admission controllers, I suggest a new flag along the lines of --enable-recommended-admission-controllers, and a warning if the existing flag is otherwise left unset.

**Comment by erictune** [comment]
I'm in favor of making this change, and having it be opt-out.

**Comment by erictune** [comment]
The release note needs to be more specific.  Maybe append something like this:


    If you previously had not set the `--admission-control` flag, your cluster behavior may change (to be more standard).  See [https://kubernetes.io/docs/admin/admission-controllers/] for explanation of admission control.

**Comment by hzxuzhonghu** [comment]
@erictune ok, will update today.

**Comment by hzxuzhonghu** [comment]
so many case failed with ServiceAccount plugin enabled.

**Comment by deads2k** [comment]
> so many case failed with ServiceAccount plugin enabled.

controllers not running maybe?

**Comment by hzxuzhonghu** [comment]
rebased

**Comment by deads2k** [comment]
Sorry, I didn't see that this got itself sorted out.

/lgtm

 squash up when you rebase

**Comment by hzxuzhonghu** [comment]
@deads2k @sttts  squashed and rebased

**Comment by hzxuzhonghu** [comment]
run gofmt

File: pkg/kubeapiserver/options/plugins.go [hunk 3]
```diff
@@ -130,8 +130,17 @@ func RegisterAllAdmissionPlugins(plugins *admission.Plugins) {
...
-	defaultOffPlugins.Delete(lifecycle.PluginName)
+	defaultOnPlugins := sets.NewString(
+		lifecycle.PluginName,                //NamespaceLifecycle
+		limitranger.PluginName,              //LimitRanger
+		serviceaccount.PluginName,           //ServiceAccount
+		label.PluginName,                    //PersistentVolumeLabel
+		setdefault.PluginName,               //DefaultStorageClass
+		defaulttolerationseconds.PluginName, //DefaultTolerationSeconds
+		mutatingwebhook.PluginName,          //MutatingAdmissionWebhook
+		validatingwebhook.PluginName,        //ValidatingAdmissionWebhook
```
Comment (lavalamp) [inline pkg/kubeapiserver/options/plugins.go]: Why are webhooks off by default?

Comment (liggitt) [reply pkg/kubeapiserver/options/plugins.go]: this looks to make them on by default, not off

Comment (lavalamp) [reply pkg/kubeapiserver/options/plugins.go]: My bad. I was reading the function name, not the variable name. Sorry for the noise.

**Pull request merged**